### DISCUSSION TITLE
How do I use DeckGL with @mapbox/mapbox-gl-directions/?

### DISCUSSION THREAD

**Discussion by TGChrisRArendt** [visgl/deck.gl#6955] [Q&A]
Hello! I am trying to figure this out for a while now :'( 

We need to add directions to our maps and found mapbox-gl-directions, which seems to cover our needs, woohoo! The problem is that I having trouble accessing the map context. Here's what I have done so far:

```
function NavigationSetup(){
  const context = React.useContext(MapContext);
  const [hasDirectionsLoaded, setHasDirectionsLoaded] = React.useState(false);

  React.useEffect(() => {
    if(!hasDirectionsLoaded){
      setHasDirectionsLoaded(true);

      context.map.addControl(
        new MapboxDirections({
          accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN
        }),
        'bottom-left'
      );
    }
  }, []);

  return <></>;
}
```

```
  const [show, setShow] = React.useState(false);
...
      <button onClick={() => setShow(true)}>show</button>
...
          <DeckGL ContextProvider={_MapContext.Provider}
                  controller={{
                    dragPan,
                    dragRotate,
                    scrollZoom
                  }}
                  getCursor={() => figureHovered ? "pointer" : getCursor()}
                  initialViewState={INITIAL_VIEW_STATE}
                  layers={[...layers, ...tempLayers]}
                  maxZoom={10}
                  minZoom={5}
                  onClick={deckGl_onClick}
                  onDblClick={onDblClick}
                  onDrag={onDrag}
                  onLoad={() => setIsMapReady(true)}
                  onViewStateChange={deckGl_onViewStateChange}
                  viewState={viewState}>
...
            <StaticMap key="map"
                       mapStyle={mapStyle}>
              <div style={{position: 'absolute', right: 30, top: 110, zIndex: 1}}>
                {show && <NavigationSetup onViewStateChange={deckGl_onViewStateChange} deckglContext={_MapContext.Provider}/>}
              </div>
            </StaticMap>
...
          </DeckGL>
```

I've cut out some stuff above because there's a bunch of other irrelevant stuff going on. 

The issues I'm having are:

1. The Deckgl layer is above the control
2. I was able to move the control using fixed position (ugh) but then StaticMap is moving independently of DeckGL and I can't figure out how I'm supposed to get / supply the correct context

I was excited when I found getMapboxMap on https://deck.gl/docs/api-reference/core/deckgl, because maybe that just already has the context figured out... but I don't know how to access this property... I tried `import { Deck } from '@deck.gl/core';` with `Deck.getMapboxMap` and `import DeckGL, { IconLayer, PathLayer } from 'deck.gl';` with `DeckGL.getMapboxMap` but it's undefined... 

I tried looking for references to getMapboxMap in the Discussion section here but I didn't see anything about it, so I thought I would open this thread. Please let me know if there's some other similar thread you know of, and thanks for taking the time to review this <3

**Comment by Pessimistress** [comment]
The documentation you linked to is about the pre-bundled version of deck.gl, not the React component. The React component has no knowledge of mapbox.

It is not possible for you to use an arbitrary mapbox control under this usage pattern. In the upcoming release we are going to publish a new class [MapboxOverlay](https://github.com/visgl/deck.gl/blob/master/docs/api-reference/mapbox/mapbox-overlay.md) that is an implementation of mapbox-gl's `IControl`. You can try it with `@deck.gl/mapbox@^8.8.0-alpha`.

**Comment by TGChrisRArendt** [comment]
I'm grateful for the quick response Pessimistress

I tried this and it works:
```
      const overlay = new MapboxOverlay({
        layers: [
          new ScatterplotLayer({
            id: 'my-scatterplot',
            data: [
              {position: [-82.9451466774461, 42.49152711639333], size: 100}
            ],
            getPosition: d => d.position,
            getRadius: d => d.size,
            getColor: [255, 0, 0]
          })
        ]
      });

      context.map.addControl(overlay);
```

However I don't understand how to add the directions to it. I tried the following:
```
      const overlay = new MapboxOverlay({
        layers: [
          new MapboxDirections({
            accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN
          })
        ]
      });
```
But I received this error/stack:
```
Uncaught TypeError: layer.raiseError is not a function
    at LayerManager._handleError (layer-manager.ts:237:1)
    at LayerManager._updateSublayersRecursively (layer-manager.ts:314:1)
    at LayerManager._updateLayers (layer-manager.ts:258:1)
    at LayerManager.setLayers (layer-manager.ts:206:1)
    at LayerManager.updateLayers (layer-manager.ts:218:1)
    at Deck._onRenderFrame (deck.js:752:1)
    at AnimationLoop.onRender (animation-loop.js:269:1)
    at AnimationLoop._renderFrame (animation-loop.js:370:1)
    at AnimationLoop.redraw (animation-loop.js:191:1)
    at renderFrame (animation-loop.js:297:1)
```

Do you think this is an issue with how I implemented it, or some bug?

**Reply by Pessimistress** [reply]
`MapboxOverlay` is a peer to the directions control:

```js
context.map.addControl(new MapboxOverlay({
	layers: [...]
}));
context.map.addControl(
        new MapboxDirections(...),
        'bottom-left'
);
```

**Comment by TGChrisRArendt** [comment]
Sorry, I'm not sure I follow. Is the idea ("mapboxoverlay is a peer") that MapboxOverlay can not help to make MapboxDirections work? Or is there a way they can be used together?

I have this code now:
```
      context.map.addControl(new MapboxOverlay({
        layers: []
      }));
      context.map.addControl(
        new MapboxDirections({
          accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN
        }),
        'bottom-left'
      );
```

I don't understand how I'm supposed to connect the two. This runs, but I still have the same issues:
1. The Deckgl layer is above the control
2. I was able to move the control using fixed position (ugh) but then StaticMap is moving independently of DeckGL and I can't figure out how I'm supposed to get / supply the correct context

**Reply by Pessimistress** [reply]
You will not use the DeckGL React component if you use the `MapboxOverlay`. In this pattern, your root component will be `Map` from react-map-gl (I do recommend you move to v7, though `InteractiveMap` from v6 should work too).

**Reply by TGChrisRArendt** [reply]
How does MapboxOverlay help to resolve the issue with using MapboxDirections? I am already able to use MapboxDirections in the same way as your above example.

**Reply by Pessimistress** [reply]
When you use this pattern:

- DeckGL
  + Map
    * MapboxDirections

DeckGL is driving the camera. Any camera movement caused by MapboxDirections is only propagated to the Map, not DeckGL, therefore they "move independently".

When you use this pattern:

- Map
  + Deck (MapboxOverlay)
  + MapboxDirections

Map is driving the camera, and they should all move together.

I suggest you create a CodeSandbox to help with the conversation.

**Reply by TGChrisRArendt** [reply]
Thanks so much for this explanation Pessimistress. I'll take a crack at it now.
